### PR TITLE
Bring back FixedDimension

### PR DESCRIPTION
**Pull request by navjotk**
Reverts opesci/devito#346

### CODE REVIEW

File: tests/test_dimension.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+@pytest.mark.xfail
+def test_incorrect_usage():
+    m = DenseData(name="m", shape=(10, 10))
+    devito.x = FixedDimension(name='x', size=10, spacing=devito.x.spacing)
+    assert(devito.x in m.indices)
+
+
+def test_correct_usage():
+    myx = FixedDimension(name='x', size=10, spacing=devito.x.spacing)
+    m = DenseData(name="m", dimensions=(myx, y))
```
Comment (mloubout) [inline tests/test_dimension.py]: `m.laplace` ?

File: tests/test_adjointA.py [hunk 2]
```diff
@@ -58,7 +59,8 @@ def test_acoustic(mkey, dimensions, time_order, space_order, nbpml, fix_dim):
                                 space_order=space_order)
 
     # Set fixed ("baked-in") time dimension if requested
-    time.size = solver.source.nt if fix_dim else None
+    if fix_dim:
+        devito.time = FixedDimension('time', size=solver.source.nt)
```
Comment (mloubout) [inline tests/test_adjointA.py]: Incorrect usage according to the test.

File: devito/dimension.py [hunk 3]
```diff
@@ -31,10 +30,38 @@ def __str__(self):
...
+
+    def __new__(cls, name, **kwargs):
+        newobj = super(FixedDimension, cls).__new__(cls, name)
+        newobj._size = kwargs.get('size', None)
+        return newobj
+
+    @property
+    def symbolic_size(self):
+        """The symbolic size of this dimension."""
+        return Number(self.size)
```
Comment (mloubout) [inline devito/dimension.py]: This is the only difference with a non-fixed dimension (unless I am missing something). Wouldn't be easier/nicer to only have a "conditional" property?

File: devito/operator.py [hunk 4]
```diff
@@ -198,7 +198,9 @@ def _dle_arguments(self, dim_sizes):
         dle_arguments = OrderedDict()
         autotune = True
         for i in self.dle_arguments:
-            dim_size = dim_sizes.get(i.original_dim.name, i.original_dim.size)
+            dim_size = dim_sizes.get(i.original_dim.name,
+                                     i.original_dim.size if i.original_dim.is_Fixed
```
Comment (FabioLuporini) [inline devito/operator.py]: Perhaps we don't need this ``if``, because if ``False`` it's gonna return ``None`` anyway?

File: devito/dimension.py [hunk 5]
```diff
@@ -31,10 +30,38 @@ def __str__(self):
     @property
     def symbolic_size(self):
         """The symbolic size of this dimension."""
-        try:
-            return Number(self.size)
-        except TypeError:
-            return self.rtargs[0].as_symbol
+        return self.rtargs[0].as_symbol
```
Comment (FabioLuporini) [inline devito/dimension.py]: Suggestion for the future: perhaps we wanna return a ``namedtuple`` when it comes to the ``Dimension``'s ``rtargs``

File: devito/operator.py (hunk 4 above)
Comment (FabioLuporini) [reply devito/operator.py]: @navjotk

**Pull request merged**